interface/ceed-basis.c

```diff
@@ -1386,6 +1383,198 @@ int CeedBasisApply(CeedBasis basis, CeedInt num_elem, CeedTransposeMode t_mode,
+    CeedCall(CeedBasisGetInterp1D(basis, &interp_1d));
+    CeedCall(CeedMatrixMatrixMultiply(basis->ceed, chebyshev_coeffs_1d, interp_1d, chebyshev_interp_1d, Q_1d, P_1d, Q_1d));
+
+    CeedCall(CeedVectorCreate(basis->ceed, num_comp * CeedIntPow(Q_1d, dim), &basis->vec_chebyshev));
+    CeedCall(CeedBasisCreateTensorH1(basis->ceed, dim, num_comp, Q_1d, Q_1d, chebyshev_interp_1d, chebyshev_grad_1d, q_ref_1d, chebyshev_q_weight_1d,
+                                     &basis->basis_chebyshev));
+
+    // Cleanup
+    CeedCall(CeedFree(&C));
+    CeedCall(CeedFree(&chebyshev_coeffs_1d));
+    CeedCall(CeedFree(&I));
+    CeedCall(CeedFree(&tau));
+    CeedCall(CeedFree(&chebyshev_interp_1d));
+    CeedCall(CeedFree(&chebyshev_grad_1d));
+    CeedCall(CeedFree(&chebyshev_q_weight_1d));
+  }
+
+  // Basis evaluation
+  switch (t_mode) {
+    case CEED_NOTRANSPOSE: {
+      // Nodes to arbitrary points
+      CeedScalar       *v_array;
+      const CeedScalar *chebyshev_coeffs, *x_array_read;
+
+      // -- Interpolate to Chebyshev coefficients
+      CeedCall(CeedBasisApply(basis->basis_chebyshev, 1, CEED_NOTRANSPOSE, CEED_EVAL_INTERP, u, basis->vec_chebyshev));
+
+      // -- Evaluate Chebyshev polynomials at arbitrary points
+      CeedCall(CeedVectorGetArrayRead(basis->vec_chebyshev, CEED_MEM_HOST, &chebyshev_coeffs));
+      CeedCall(CeedVectorGetArrayRead(x_ref, CEED_MEM_HOST, &x_array_read));
+      CeedCall(CeedVectorGetArrayWrite(v, CEED_MEM_HOST, &v_array));
+      {
+        CeedScalar tmp[2][num_comp * CeedIntPow(Q_1d, dim)], chebyshev_x[Q_1d];
+
+        // ---- Clear output array
+        for (CeedInt i = 0; i < num_points * num_comp; i++) v_array[i] = 0.0;
+        // ---- Values at point
+        for (CeedInt p = 0; p < num_points; p++) {
+          CeedInt pre = num_comp * CeedIntPow(Q_1d, dim - 1), post = 1;
+          for (CeedInt d = 0; d < dim; d++) {
+            // ------ Compute Chebyshev polynomial values
+            {
+              const CeedScalar x = x_array_read[p * dim + d];
+
+              chebyshev_x[0] = 1.0;
+              chebyshev_x[1] = 2 * x;
+              for (CeedInt j = 2; j < Q_1d; j++) {
+                chebyshev_x[j] = 2 * x * chebyshev_x[j - 1] - chebyshev_x[j - 2];
+              }
+            }
+            // ------ Tensor contract
+            {
+              const CeedScalar *in  = d == 0 ? chebyshev_coeffs : tmp[d % 2];
+              CeedScalar       *out = d == dim - 1 ? &v_array[p * num_comp] : tmp[(d + 1) % 2];
+
+              for (CeedInt q = 0; q < pre * post; q++) out[q] = (CeedScalar)0.0;
+              for (CeedInt a = 0; a < pre; a++) {
+                for (CeedInt b = 0; b < Q_1d; b++) {
+                  for (CeedInt c = 0; c < post; c++) out[a * post + c] += chebyshev_x[b] * in[(a * Q_1d + b) * post + c];
+                }
+              }
+            }
+            pre /= Q_1d;
+            post *= 1;
+          }
+        }
+      }
```

Comment:
We can presumably call existing `TensorContract` here. I assume this will be the "hot" spot we'll seek to optimize eventually.

Reply:
For the CPU, yeah, I plan on using the CeedBasis's CeedTensorContract object. On any backend that doesn't have a CeedTensorContract object, I think I'll make a `/cpu/self` ceed and then use that to make a CeedTensorContract (until we get GPU impls)

Reply:
See the latest push (also using the `TensorContract` object makes clang-tidy happier about buffers it couldn't verify were clean, which probably should give me some pause about clang-tidy's ability to reason through function calls)